File: Group_the_People_Given_the_Group_Size They_Belong_To.py
class Solution:
    def groupThePeople(self, groupSizes: List[int]) -> List[List[int]]:
        res = []
        groups = defaultdict(list)

        for i, size in enumerate(groupSizes):
            groups[size].append(i)

            if len(groups[size]) == size:
                res.append(groups.pop(size))
        
        return res

File: Group_the_People_Given_the_Group_Size They_Belong_To.js
/**
 * @param {number[]} groupSizes
 * @return {number[][]}
 */
var groupThePeople = function(groupSizes) {
    const res = [];
    const groups = {};

    for (let i = 0; i < groupSizes.length; i++) {
        const size = groupSizes[i];

        if (!groups[size]) {
            groups[size] = [];
        }

        groups[size].push(i);

        if (groups[size].length === size) {
            res.push(groups[size]);
            delete groups[size];
        }
    }

    return res;    
};
